### PR TITLE
Update ESA-YOLO-Plus: Added SPD, TBD, CPA modules

### PR DESCRIPTION
**Pull request by vinhisreal**
## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
Adds three new ESA-YOLO-Plus building blocks (SPDConv, TBD, CPA) and wires them into the Ultralytics module registry and model parser so they can be used in YAML model definitions. 

### 📊 Key Changes
- ➕ Introduces `SPDConv` (space-to-depth convolution) to preserve fine-grained spatial information while increasing channel depth.
- ➕ Adds `TBD`, a tri-branch downsampling module combining max-pooling, average-pooling, and strided convolution with fusion.
- ➕ Implements `CPA`, a dual-branch attention block combining channel-wise (color) and spatial (position) attention.
- 🔗 Registers `SPDConv`, `TBD`, and `CPA` in `ultralytics.nn.modules.__init__`, `block.py`, and `tasks.py` so they are available in ESA-YOLO-Plus and other YAML-configurable models via `parse_model`.

### 🎯 Purpose & Impact
- 🚀 Extends ESA-YOLO-Plus with richer backbone/neck components that can better capture fine details and multi-scale context.
- 🎨 Improves feature expressiveness through combined channel and spatial attention, potentially boosting detection accuracy on complex scenes.
- 🧩 Keeps the design modular: the new blocks are now first-class modules that can be reused or experimented with in custom YOLO configurations.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @vinhisreal, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! This is an automated review helper to get things started; a Ultralytics engineer will be by shortly to provide a detailed review.

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_  — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

File: ultralytics/nn/modules/block.py [hunk 1]
```diff
@@ -1943,3 +1946,63 @@ def forward(self, x: list[torch.Tensor], vp: torch.Tensor) -> torch.Tensor:
...
+
+class SPDConv(nn.Module):
+    # Space-to-Depth Convolution để bảo toàn thông tin hạt mịn
+    def __init__(self, inc, outc, dimension=1):
+        super().__init__()
+        self.d = 2
+        # Sau khi space-to-depth, số kênh tăng gấp 4 (c1 * 4)
+        self.conv = Conv(inc * 4, outc, 3, 1) # Conv tiêu chuẩn của Ultralytics
+
+    def forward(self, x):
```
Comment (UltralyticsAssistant) [inline ultralytics/nn/modules/block.py:1958]: ⚠️ **HIGH**: `SPDConv.forward()` slices every other pixel and concatenates the four quadrants, but when `H` or `W` is odd the slices coming from `x[..., 1::2, :]` are one element smaller, so `torch.cat` raises a size mismatch. Either pad/crop before slicing or use a dedicated space-to-depth op that tolerates odd spatial dimensions.

**Suggested change:**
```suggestion
    x = F.pad(x, (0, x.shape[-1] % 2, 0, x.shape[-2] % 2))
```

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Address the SPDConv odd-dimension crash, remove or use the unused `dimension` argument, and guard CPA’s reduction ratio to avoid zero-width linear layers.

💬 Posted 3 inline comments

File: ultralytics/nn/modules/block.py [hunk 2]
```diff
@@ -1943,3 +1946,63 @@ def forward(self, x: list[torch.Tensor], vp: torch.Tensor) -> torch.Tensor:
...
+        # Hợp nhất 3 nhánh: c1 (max) + c1 (avg) + c1 (conv) -> c2
+        self.fusion = Conv(c1 * 3, c2, 1, 1)
+
+    def forward(self, x):
+        x_max = self.maxpool(x)
+        x_avg = self.avgpool(x)
+        x_conv = self.conv(x)
+        return self.fusion(torch.cat([x_max, x_avg, x_conv], 1))
+class CPA(nn.Module):
+    def __init__(self, c1, c2, reduction=16):
```
Comment (UltralyticsAssistant) [inline ultralytics/nn/modules/block.py:1981]: 💡 **MEDIUM**: `c1 // reduction` can be zero when `c1 < reduction`, causing `nn.Linear` to instantiate with zero out_features and crash at runtime. Clamp the reduced width to at least 1 before building the MLP so small channel counts remain valid.

**Suggested change:**
```suggestion
    hidden = max(1, c1 // reduction)
```

File: ultralytics/nn/modules/block.py [hunk 3]
```diff
@@ -1943,3 +1946,63 @@ def forward(self, x: list[torch.Tensor], vp: torch.Tensor) -> torch.Tensor:
         aggregated = score.transpose(-2, -3) @ x.reshape(B, self.c, C // self.c, -1).transpose(-1, -2)
 
         return F.normalize(aggregated.transpose(-2, -3).reshape(B, Q, -1), dim=-1, p=2)
+
+class SPDConv(nn.Module):
+    # Space-to-Depth Convolution để bảo toàn thông tin hạt mịn
+    def __init__(self, inc, outc, dimension=1):
```
Comment (UltralyticsAssistant) [inline ultralytics/nn/modules/block.py:1952]: 📝 **LOW**: `dimension` is exposed as an argument but never used; the module hardcodes `self.d = 2`, so callers cannot configure or even benefit from it. Drop the unused parameter or wire it into the space-to-depth logic to avoid a misleading API.

**Suggested change:**
```suggestion
    def __init__(self, inc, outc, dimension=2):
```
